src/platform/linux/mod.rs

```diff
@@ -386,76 +390,84 @@ impl OsIpcChannel {
 }
 
 pub struct OsIpcReceiverSet {
-    pollfds: Vec<pollfd>,
-}
-
-impl Drop for OsIpcReceiverSet {
-    fn drop(&mut self) {
-        unsafe {
-            for pollfd in self.pollfds.iter() {
-                let result = libc::close(pollfd.fd);
-                assert!(thread::panicking() || result == 0);
-            }
-        }
-    }
+    poll: Poll,
+    pollfds: HashMap<Token, c_int, BuildHasherDefault<FnvHasher>>,
+    events: Events
 }
 
 impl OsIpcReceiverSet {
     pub fn new() -> Result<OsIpcReceiverSet,UnixError> {
+        let fnv = BuildHasherDefault::<FnvHasher>::default();
         Ok(OsIpcReceiverSet {
-            pollfds: Vec::new(),
+            poll: Poll::new().expect("Could not create a new receiver"),
+            pollfds: HashMap::with_hasher(fnv),
+            events: Events::with_capacity(1024)
         })
     }
 
     pub fn add(&mut self, receiver: OsIpcReceiver) -> Result<i64,UnixError> {
         let fd = receiver.consume_fd();
-        self.pollfds.push(pollfd {
-            fd: fd,
-            events: POLLIN,
-            revents: 0,
-        });
+        let io = EventedFd(&fd);
+        let fd_token = Token(fd as usize);
+        self.poll.register(&io, fd_token, Ready::readable(), PollOpt::level())
+            .expect(&format!("Could not register: {:?}", fd));
+        self.pollfds.insert(fd_token, fd);
         Ok(fd as i64)
     }
 
     pub fn select(&mut self) -> Result<Vec<OsIpcSelectionResult>,UnixError> {
         let mut selection_results = Vec::new();
-        let result = unsafe {
-            poll(self.pollfds.as_mut_ptr(), self.pollfds.len() as nfds_t, -1)
+        let sz = match self.poll.poll(&mut self.events, None) {
+            Ok(sz) if sz > 0 => sz,
+            _ => { return Err(UnixError::last()); }
         };
-        if result <= 0 {
-            return Err(UnixError::last())
-        }
 
         let mut hangups = HashSet::new();
-        for pollfd in self.pollfds.iter_mut() {
-            if (pollfd.revents & POLLIN) != 0 {
-                match recv(pollfd.fd, BlockingMode::Blocking) {
-                    Ok((data, channels, shared_memory_regions)) => {
-                        selection_results.push(OsIpcSelectionResult::DataReceived(
-                                pollfd.fd as i64,
-                                data,
-                                channels,
-                                shared_memory_regions));
-                    }
-                    Err(err) if err.channel_is_closed() => {
-                        hangups.insert(pollfd.fd);
-                        selection_results.push(OsIpcSelectionResult::ChannelClosed(
-                                    pollfd.fd as i64))
+        for i in 0..sz {
+            let evt = self.events.get(i).unwrap();
```

Comment:
At the time of writing this I think events didn't impl Iterator? maybe? I know there was some reason why I did this, but yes. Now this is possible.